### ISSUE TITLE
mysql go client autodetection_retry doesn't seem to work

### ISSUE THREAD

**Issue by lucasRolff** [netdata/netdata#11131]
##### Bug report summary

I use Netdata extensively and recently made the switch from the python MySQL client to the Golang MySQL client in netdata.

However, whenever MySQL restarts (during upgrade or maintenance), it quite often results in the monitoring of MySQL stops (because of no metrics).

I've configured `autodetection_retry` to 60 in the `mysql.conf`, sadly I never see retry messages in the log, and monitoring of MySQL stays down for hours.

##### OS / Environment

```
/etc/cloudlinux-release:CloudLinux release 8.3 (Pavel Belyayev)
/etc/os-release:NAME="CloudLinux"
/etc/os-release:VERSION="8.3 (Pavel Belyayev)"
/etc/os-release:ID="cloudlinux"
/etc/os-release:ID_LIKE="rhel fedora centos"
/etc/os-release:VERSION_ID="8.3"
/etc/os-release:PLATFORM_ID="platform:el8"
/etc/os-release:PRETTY_NAME="CloudLinux 8.3 (Pavel Belyayev)"
/etc/os-release:ANSI_COLOR="0;31"
/etc/os-release:CPE_NAME="cpe:/o:cloudlinux:cloudlinux:8.3:GA:server"
/etc/os-release:HOME_URL="https://www.cloudlinux.com/"
/etc/os-release:BUG_REPORT_URL="https://www.cloudlinux.com/support"
/etc/redhat-release:CloudLinux release 8.3 (Pavel Belyayev)
/etc/system-release:CloudLinux release 8.3 (Pavel Belyayev)
```

##### Netdata version

```
Version: netdata v1.30.1
Configure options:  '--prefix=/usr' '--sysconfdir=/etc' '--localstatedir=/var' '--libexecdir=/usr/libexec' '--libdir=/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--disable-ebpf' '--enable-plugin-freeipmi' '--with-bundled-lws=externaldeps/libwebsockets' '--with-libJudy=externaldeps/libJudy' 'CFLAGS=-O2' 'LDFLAGS='
Features:
    dbengine:                YES
    Native HTTPS:            YES
    Netdata Cloud:           YES
    Cloud Implementation:    Legacy
    TLS Host Verification:   YES
Libraries:
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  NO
    libcrypto:               YES
    libm:                    YES
    LWS:                     YES static v3.2.2
    mosquitto:               YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    NO
    EBPF:                    NO
    IPMI:                    YES
    NFACCT:                  NO
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: NO
```

##### Installation method

```
bash <(curl -Ss https://my-netdata.io/kickstart.sh)
```

##### Component Name

MySQL Golang

##### Steps To Reproduce

1. Configure autodetection_retry to 60 seconds (or lower)
2. Kill MySQL after monitoring has started, wait a bit, start MySQL again
3. Netdata won't pick up monitoring MySQL again

##### Expected behavior

The retry mechanism should work.

##### Additional info

config:
```
# netdata go.d.plugin configuration for mysql
#
# This file is in YAML format. Generally the format is:
#
# name: value
#
# There are 2 sections:
#  - GLOBAL
#  - JOBS
#
#
# [ GLOBAL ]
# These variables set the defaults for all JOBs, however each JOB may define its own, overriding the defaults.
#
# The GLOBAL section format:
# param1: value1
# param2: value2
#
# Currently supported global parameters:
#  - update_every
#    Data collection frequency in seconds. Default: 1.
#
#  - autodetection_retry
#    Re-check interval in seconds. Attempts to start the job are made once every interval.
#    Zero means not to schedule re-check. Default: 0.
#
#  - priority
#    Priority is the relative priority of the charts as rendered on the web page,
#    lower numbers make the charts appear before the ones with higher numbers. Default: 70000.
#
#
# [ JOBS ]
# JOBS allow you to collect values from multiple sources.
# Each source will have its own set of charts.
#
# IMPORTANT:
#  - Parameter 'name' is mandatory.
#  - Jobs with the same name are mutually exclusive. Only one of them will be allowed running at any time.
#
# This allows autodetection to try several alternatives and pick the one that works.
# Any number of jobs is supported.
#
# The JOBS section format:
#
# jobs:
#   - name: job1
#     param1: value1
#     param2: value2
#
#   - name: job2
#     param1: value1
#     param2: value2
#
#   - name: job2
#     param1: value1
#
#
# [ List of JOB specific parameters ]:
#  - dsn
#    Data Source Name.
#    Format: [username[:password]@][protocol[(address)]]/dbname[?param1=value1&...&paramN=valueN]
#    Except for the databasename, all values are optional. So the minimal DSN is: /dbname
#    If you do not want to preselect a database, leave dbname empty: /
#    Full description: https://github.com/go-sql-driver/mysql#dsn-data-source-name
#    Syntax:
#     dsn: netdata@tcp(127.0.0.1:3306)/
#
#  - my.cnf
#    Specifies my.cnf file to read connection parameters from under the [client] section
#    Syntax:
#     my.cnf: '/etc/my.cnf'
#
#
# [ JOB defaults ]:
#  No parameters
#
#
# [ JOB mandatory parameters ]:
#  - name
#  - dsn
# ------------------------------------------------MODULE-CONFIGURATION--------------------------------------------------
# [ GLOBAL ]
# update_every: 1
autodetection_retry: 60
# priority: 70000
#
#
# [ JOBS ]
jobs:
  # netdata
  - name: local
    dsn: netdata@unix(/var/run/mysqld/mysqld.sock)/

  - name: local
    dsn: netdata@unix(/var/run/mysqld/mysql.sock)/

  - name: local
    dsn: netdata@unix(/var/lib/mysql/mysql.sock)/

  - name: local
    dsn: netdata@unix(/tmp/mysql.sock)/

  - name: local
    dsn: netdata@tcp(127.0.0.1:3306)/

  - name: local
    dsn: netdata@tcp([::1]:3306)/
```

**Comment by ilyam8** [comment]
Hi @lucasRolff 👋 

> Kill MySQL after monitoring has started, wait a bit, start MySQL again

This should work w/o the `autodetection_retry` feature, the feature supposed to help you when the service starts after Netdata.

Could you do the following:
 - (assuming you have MySQL running)
 - run the go.d.plugin mysql module in the debug mode, ensure it connect to the MySQL instance and returns data
 - stop MySQL; wait a bit; start MySQL
 - see the debug output, copy it and post it here. mysql collector supposed to reconnect and continue collect data.

**Comment by ilyam8** [comment]
Ok, I did some tests using [bitnami/bitnami-docker-mariadb-galera](https://github.com/bitnami/bitnami-docker-mariadb-galera) and docker stop/start

The gaps are docker stop mariadb; wait; docker start mariadb ():

<img width="1138" alt="Screenshot 2021-05-11 at 10 48 09" src="https://user-images.githubusercontent.com/22274335/117778586-77a9be80-b246-11eb-9790-957069ed7974.png">

---

The is [some penalty value for failing jobs](https://github.com/netdata/go.d.plugin/blob/fb4772c13c3beba5b8d5229dfefe8f62a64c7f6a/agent/module/job.go#L443-L449), it slowly increases every 5 fails in a row, so it could take some time if your MySQL instance wasn't running for a long time.

```go
func (j Job) penalty() int {
	v := j.retries / penaltyStep * penaltyStep * j.updateEvery / 2
	if v > maxPenalty {
		return maxPenalty
	}
	return v
}
```

**Comment by ilyam8** [comment]
> It seems if it's used socket-based connections (such as /var/lib/mysql/mysql.sock), then it terminates, if it's trying to use TCP (127.0.0.1:3306) - then it will actually end up reconnecting when MySQL comes online again

Nice finding! Indeed, i was using `TCP` socket in my test, not `unix` socket.

**Comment by ilyam8** [comment]
> With that said, even with autodetection_retry: 5 set, and TCP connections, I'm now at tick 120, and still haven't tried to connect, if MySQL were down prior to Netdata starting collecting metrics.

Confirm this.

That is how we _init_ and _check_ the collector:
 - Init(): always once, if it fails we drop the data collection job.
 - Check(): once or every `autodetection_retry` if configured.
 - Collect(): every `update_every`
 
The problem is we [open a connection in Init()](https://github.com/netdata/go.d.plugin/blob/fb4772c13c3beba5b8d5229dfefe8f62a64c7f6a/modules/mysql/mysql.go#L77-L95) and it fails if MySQL is not running. And we never retry because `Init()` supposed to be called only once.

**Comment by ilyam8** [comment]
@lucasRolff the _autodetection_retry doesn't seem to work_ bug is fixed in https://github.com/netdata/go.d.plugin/pull/577

The problem with UNIX socket is another problem and I am going to reproduce it and create a separate issue.

This one will be closed when the fix is merged.
